A few improvements regarding the dependency on the operating system language & Some metric output "improvements"

Seji64 opened the pull request:
## Pull Request Checklist

**Is this in reference to an existing issue?**

Yes see #16 

#### General

- [ ] Update Changelog following the conventions laid out on [Keep A Changelog](http://keepachangelog.com/)

- [ ] Update README with any necessary configuration snippets

- [ ] Binstubs are created if needed

- [ ] RuboCop passes

- [x] Existing tests pass

#### New Plugins

- [ ] Tests

- [ ] Add the plugin to the README

- [ ] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

#### Purpose

Allows testing of CPU load independent of operating system language

#### Known Compatibility Issues

None

majormoses commented:
@simonsteur any chance you can review and test this? More than happy to merge and release I just don't really know powershell all that well.

simonsteur commented:
@majormoses no problem, I'll do it as soon as I get a chance.

majormoses commented on bin/powershell/metric-windows-cpu-load.ps1:
```diff
@@ -1,18 +1,4 @@
 #
-#   metric-windows-cpu-load.ps1
```
why are we removing this?

Seji64 replied:
To be honest, I don't know why the header is removed in the change. Wasn't and isn't my intention.

majormoses commented on bin/powershell/perfhelper.ps1:
```diff
@@ -0,0 +1,81 @@
+##################################################
```
Lets include the standard header (you can still have this below): https://github.com/sensu-plugins/documentation/blob/master/docs/developer_guidelines.md#coding-style you can tweak it to be less ruby specific of course.

Seji64 replied:
Sure, no problem. I will have at look at it tomorrow

Seji64 commented:
I added the Headers as requested, hope it fits

majormoses commented on bin/powershell/metric-windows-cpu-load.ps1:
```diff
@@ -1,35 +1,61 @@
 #
-#   metric-windows-cpu-load.ps1
+#   check-windows-cpu-load.ps1
```
`metric-windows-cpu-load.ps1`

majormoses commented on bin/powershell/metric-windows-cpu-load.ps1:
```diff
@@ -1,35 +1,61 @@
 #
 # DESCRIPTION:
-#   This plugin collects and outputs the CPU Usage in a Graphite acceptable format.
+#   This plugin collects the CPU Usage and compares against the WARNING and CRITICAL thresholds.
```
did this really change?

majormoses commented on bin/powershell/metric-windows-cpu-load.ps1:
```diff
@@ -1,35 +1,61 @@
 #
 # DEPENDENCIES:
-#   Powershell
+#   Powershell 3.0 or above
```
:+1:

majormoses commented on bin/powershell/metric-windows-cpu-load.ps1:
```diff
@@ -1,35 +1,61 @@
 #
 # USAGE:
-#   Powershell.exe -NonInteractive -NoProfile -ExecutionPolicy Bypass -NoLogo -File C:\\etc\\sensu\\plugins\\metric-windows-cpu-load.ps1
+#   Powershell.exe -NonInteractive -NoProfile -ExecutionPolicy Bypass -NoLogo -File C:\\etc\\sensu\\plugins\\check-windows-cpu-load.ps1 90 95
```
`metric`

simonsteur commented on bin/powershell/metric-windows-cpu-load.ps1:
```diff
@@ -22,14 +22,40 @@
+$Path = [System.Net.Dns]::GetHostEntry([string]"localhost").HostName.toLower()
 
-$Time = [int][double]::Parse((Get-Date -UFormat %s))
+$Time = DateTimeToUnixTimestamp -DateTime (Get-Date)
```
```
-$Path = hostname
-$Path = $Path.ToLower()
-$Value = (Get-WmiObject CIM_Processor).LoadPercentage
 +$Path = [System.Net.Dns]::GetHostEntry([string]"localhost").HostName.toLower()
-$Time = [int][double]::Parse((Get-Date -UFormat %s))
 +$Time = DateTimeToUnixTimestamp -DateTime (Get-Date)
```


This doesn't seem like much of an improvement to me. What where simple oneliners became complicated ones or in the time case even a dependency on another ps1 file. And it does exactly the same in the end?

simonsteur replied:
Same goes for every other file where this same thing happens. Seems like over-complicating something super simple to me?

Seji64 replied:
Hi simon,

> Have you tested this on multiple different languages btw?

i just tested it against a german and an english windows server 2016.

> This doesn't seem like much of an improvement to me. What where simple oneliners became complicated ones or in the time case even a dependency on another ps1 file. And it does exactly the same in the end?

InfluxDB or Grafana expects the time in UTC format. If you have a german based windows installation the timecode ist just wrong and you don't get any data to see in grafana. That's why i am implemented this function. 

```
Function DateTimeToUnixTimestamp([datetime]$DateTime)
{
    $utcDate = $DateTime.ToUniversalTime()
    # Convert to a Unix time without any rounding
    [uint64]$UnixTime = [double]::Parse((Get-Date -Date $utcDate -UFormat %s))
    return [uint64]$UnixTime
}

Write-Host "Current local Time:"
Get-Date

Write-Host "Time with new function:"
DateTimeToUnixTimestamp -DateTime (Get-Date)

Write-Host "Old variant:"
[int][double]::Parse((Get-Date -UFormat %s))
```

Output:
```
Current local Time:

Freitag, 1. September 2017 13:40:05
Time with new function:
1504266006
Old variant:
1504273206
```

simonsteur replied:
Hmmz, I thought they would result in the same but apparently not. I must also admit that I've never done it that way myself either. Wouldn't replacing it with a one liner like the one below be "better" than loading in an additional file? 

`$Time = [int64](([datetime]::UtcNow)-(get-date "1/1/1970")).TotalSeconds`

Granted that I've not looked at the performance impact of this, if any at all. Just trying to keep it as simple and lightweight as possible.

Seji64 replied:
the external file must be loaded anyway. So I find it makes no difference if you use the function or not. Of course, you can write the function in every file separately, but I personally consider it unnecessary.

simonsteur replied:
only in the ones where you're looking at performance counters, but granted it's better to have a unified way then making a couple of exceptions.

I still don't  understand this though 
`$Path = [System.Net.Dns]::GetHostEntry([string]"localhost").HostName.toLower()` 

you can do the same with `$Path = ($env:computername).ToLower()` which excludes the domain name, which in my mind would only muddle the metric path?

Other than that everything looks fine to me, I don't have a way to test this against other languages myself however so gonna have to trust you on that it actually works as advertised :)

Seji64 replied:
Oh that is just a personal think, I would would like to have the fqdn in the metrics. but we can revert that cause not everybody want to have it like this :-)

majormoses replied:
sounds like an option can toggle this behavior

majormoses commented on bin/powershell/metric-windows-cpu-load.ps1:
```diff
@@ -23,45 +23,42 @@
-$ThisProcess.PriorityClass = "BelowNormal"
+if ($UseFullyQualifiedHostname -eq $false) {
+    $Path = ($env:computername).ToLower()
+}else{
```
apologies if that is how people in powershell as I know almost nothing about it. My developer senses are tingling and telling me that there either should be a space between `}else` if not a newline.

majormoses commented:
@Seji64 is it ready for another round of reviews from @simonsteur ?

Seji64 commented:
@majormoses Sure :-)

simonsteur commented:
@majormoses @Seji64 looks fine to me now. Don't have the full ability to test on the different languages but everything looks in order for the rest :)

majormoses commented:
@Seji64 I forgot to have you add a CHANGELOG entry I added one here: https://github.com/sensu-plugins/sensu-plugins-windows/commit/8d3c6041d16e71776bb43c0db3f3e8997108da42 can you please let me know if there is anything else I missed?

majormoses commented:
released: https://rubygems.org/gems/sensu-plugins-windows/versions/2.2.0

Seji64 commented:
@majormoses 

- I also added added some system related metrics (see _metric-windows-system.ps1_)

- With regard to the network check, a few hints should be packed into the README file. You **can** specify one or more interfaces (but is not obligatory). If the specified interfaces have blanks in their names,  they must be replaced with underlined in the Sensu Check. The Powershell script then converts them into spaces. For example:
`
powershell.exe -NonInteractive -NoProfile -ExecutionPolicy Bypass -NoLogo -File C:\\etc\\sensu\\plugins\\metric-windows-network.ps1 -Interfaces red_hat_virtio_ethernet_adapter
`
- Fixed wrong Timestamp 'calculation' ( DateTimeToUnixTimestamp )

That should be all - Thanks.

majormoses commented:
@Seji64 can you create a PR to add the appropriate changelog entries?

Koxx3 commented:
Hello,
I still have issue with french.

example : 
Powershell.exe -NonInteractive -NoProfile -ExecutionPolicy Bypass -NoLogo -File C:/opt/sensu/sensu-plugins-windows/bin/powershell/check-windows-ram.ps1 -c 1 -w 2

>> Get-WmiObject : Le service sp�cifi� n'existe pas en tant que service install�. 
(Exception de HRESULT : 0x80070424)

majormoses commented:
@Koxx3 I have reopened #16 let's take the discussion there.
